Introduce "Core" framework for Jetpack Stats Widget as Swift Package

mokagio opened the pull request:
Same as https://github.com/wordpress-mobile/WordPress-iOS/pull/21962 but as a Swift Package, to compare approaches.

Hat tip @kean for suggesting this comparison https://github.com/wordpress-mobile/WordPress-iOS/pull/21962#issuecomment-1799337944.

Screenshot of my smoke test to ensure that the Stats Widgets still run:

<table>
<tr>
 <td>

![Simulator Screenshot - iPhone 15 Pro - 2023-11-17 at 16 25 00](https://github.com/wordpress-mobile/WordPress-iOS/assets/1218433/c61db8b2-eba3-4ded-aece-abecf1588ffa)

</td>
 <td>

![Simulator Screenshot - iPhone 15 Pro - 2023-11-17 at 16 24 46](https://github.com/wordpress-mobile/WordPress-iOS/assets/1218433/058a9b14-a0a6-496e-a3ee-2be2131248dc)

</td>
</table>

---

WIP note: This is not done yet, but I'm at EOD and wanted to link this seedling to the classic framework PR.

mokagio committed:
Add Swift Package shell for JetpackStatsWidgetsCore

mokagio committed:
Make dummy test pass

mokagio committed:
Copy `WidgetUrlSource` and its tests in JetpackStatsWidgetsCore

We'll use this code in the app extension next. Here, I just wanted to
see the tests pass.

mokagio commented on JetpackStatsWidgetsCore/.swiftpm/xcode/xcshareddata/xcschemes/JetpackStatsWidgetsCoreTests.xcscheme:
```diff
@@ -0,0 +1,53 @@
+      buildConfiguration = "Release"
+      revealArchiveInOrganizer = "YES">
+   </ArchiveAction>
+</Scheme>
```
Note that this was not created automatically. Like all other schemes Xcode make, one has to uncheck and recheck the "Shared" checkbox in the scheme settings to move it into the `xcshareddata` folder.

mokagio commented on JetpackStatsWidgetsCore/Package.swift:
```diff
@@ -0,0 +1,19 @@
+        .target(name: librarySourceName),
+        .testTarget(
+            name: "\(librarySourceName)Tests",
+            dependencies: [.target(name: librarySourceName)]
```
The template for the packages uses

```swift
            dependencies: ["MyLibrary"]
```

But if you replace `"MyLibrary"` with the name of the library, the code no longer compiles because the expected type is `[Target.Depency]`, not `[String]`.

mokagio commented on WordPress/WordPress.xcodeproj/project.pbxproj:
```diff
@@ -10753,7 +10753,7 @@
 			sourceTree = "<group>";
 		};
-		29B97314FDCFA39411CA2CEA = {
+		29B97314FDCFA39411CA2CEA /* CustomTemplate */ = {
```
Xcode and CocoaPods keep changing this and the other lines unrelated to the package in this diff. I don't know why... One sets it one way, the other reverts.

mokagio committed:
Use JetpackStatsWidgetsCore framework in Jetpack widgets extension

Also replace existing test for the `WidgetUrlSource` that added the file
to the WordPress target in order to test it with one that imports the
package in the test target (no overhead in the app) to access the type
that is being used.

This is still boardline inappropriate, given that the unit tests for the
app target should cover the app target behavior, but it's the best we
can do without further extracting concepts such as `DeepLinkSource` in a
dedicated library.

mokagio committed:
Run JetpackStatsWidgetCore unit tests as part of WordPress unit tests

This way, we always ensure we're covered. In particular, this is the
test plan that runs in CI.

mokagio commented on WordPress/WordPressTest/WordPressUnitTests.xctestplan:
```diff
@@ -36,6 +36,13 @@
+        "identifier" : "JetpackStatsWidgetsCoreTests",
+        "name" : "JetpackStatsWidgetsCoreTests"
+      }
+    },
```
For a target this small (one test class, two tests), it's not worth spinning up a dedicated VM in CI to run its tests.

We can get the same coverage by running the tests as part of the unit tests plan. 

_If_ we'll begin adding more and more modules, we might decide to have a "all unit tests" plan for CI and one "only app unit tests" for local development. That would come with the overhead that you'd have to _remember_ to switch scheme/test-plan depending on the module you are working in.

mokagio committed:
Use swift-tools 5.8 instead of 5.9 because that's what CI has

mokagio committed:
Fix typo in JetpackStatsWidgetsCore tests folder name

mokagio committed:
Remove shared scheme for JetpackStatsWidgetsCore

It made Xcode think the package was not a package after switching
branches and broke the build.

staskus commented:
Appreciate the work!

I agree with @kean sentiment on framework target vs Swift package and would give this solution a shot 👍

mokagio commented:
Thanks @staskus ! I'd love to hear what if our trusted @crazytonyli has some comment or concern. Then, ready to see how this plays out.

crazytonyli commented:
Looks good to me. What do you think turning this swift package into an "umbrella" one, which can have many products? At the moment, there is only "JetpackStatsWidgetCore". But later, we can add more libraries, maybe "ShareExtensionCore" or something, without the need to create a new swift package and redo the Xcode project integration.

kean commented:
The app compiles and runs, and I'm able to open the package separately and run tests for it separately. I'm even able to run the tests for the JetpackStatsWidgetCore library from the Xcode workspace  👍

> What do you think turning this swift package into an "umbrella" one, which can have many products?

I agree with the Tony's suggestions. Examples: [Vapor](https://github.com/vapor/vapor/blob/main/Package.swift), [pointerfree/isowords](https://github.com/pointfreeco/isowords/blob/main/Package.swift). The example of the setup I would rather avoid: [IceCubes](https://github.com/Dimillian/IceCubesApp), where every module has its own Package.swift file.

<img width="319" alt="Screenshot 2023-11-15 at 3 57 27 PM" src="https://github.com/wordpress-mobile/WordPress-iOS/assets/1567433/c7cab8d0-6cc4-4f6a-b157-8894bc996761">

The package should also be moved down in the folder hierarchy. I expect us to create at least a dozen of these, so I'll become unmergeable if they are in the root folder.

mokagio committed:
Rename JetpackStatsWidgetsCore package to Modules

The idea is to have one single package with many targets, so that we
don't need to repeat the integration step

Co-authored-by: Tony Li <tony.li@automattic.com>

See
https://github.com/wordpress-mobile/WordPress-iOS/pull/22000#issuecomment-1809414106

mokagio commented:
@kean @crazytonyli thank you for the suggestions of having one umbrella package and to avoid multiple subpackages folder in the projects root.

I renamed `<root>/JetpackStatsWidgetCore` to `<root>/Modules` to signal that there will be more sub-packages in the future. I like "module" over "frameworks" or "libraries" because module, to me, represent a self-contained component of an architecture that can be plugged-in and hot-swapped. That's not to say that I'm opposed to extra small sharp libraries into packages, but I think the biggest ROI when it comes to deconstructing the apps architecture will be in defining modules.

Renaming is easy. Let me know if the name doesn't work out as well as I think.

I have some thoughts on how to organize the `Modules/` folder internally, but they are unrelated to this first setup effort. See https://github.com/wordpress-mobile/WordPress-iOS/pull/22047/

mokagio commented:
@wpmobilebot rightfully points out that there are no screenshots despite a view file change.

Here they are (will also update the description):

<table>
<tr>
 <td>

![Simulator Screenshot - iPhone 15 Pro - 2023-11-17 at 16 25 00](https://github.com/wordpress-mobile/WordPress-iOS/assets/1218433/c61db8b2-eba3-4ded-aece-abecf1588ffa)

</td>
 <td>

![Simulator Screenshot - iPhone 15 Pro - 2023-11-17 at 16 24 46](https://github.com/wordpress-mobile/WordPress-iOS/assets/1218433/058a9b14-a0a6-496e-a3ee-2be2131248dc)

</td>
</table>

mokagio commented:
I got some feedback from @kean on #22047 but nothing about the folder name change to `Modules/`. As mentioned https://github.com/wordpress-mobile/WordPress-iOS/pull/22000#issuecomment-1813648712 renaming is straightforward. So, in the interest of moving along with moving the unit tests to the Jetpack host app, I'll merge this.
